processors/base/helpers/__init__.py

```diff
@@ -271,6 +270,54 @@ def find_trial_by_identifiers(conn, identifiers, ignore_record_id=None):
     return trial
 
 
+def find_trial(conn, trial, source_id, record_id=None):
+    """Find if the trial already exists. Firstly, via identifiers. Secondly, via public title
+    Args:
+        conn (dict): connection dict
+        trial (dict): trial that may be found
+        source_id (text): trial's source
+        record_id (str): record id from warehouse
+    Returns:
+        dict: trial
+    """
+
+    def write_log(trial_id, record_id, method):
+        log = conn['database']['trial_deduplication_logs'].find_one(
+            trial_id=trial_id, record_id=record_id, method=method)
+
+        if not log:
+            repo = git.Repo(search_parent_directories=True)
+            commit_sha = repo.head.object.hexsha
+
+            data = {
+                'trial_id': trial_id,
+                'record_id': record_id,
+                'method': method,
+                'commit': commit_sha
+            }
+            conn['database']['trial_deduplication_logs'].insert(data)
+
+    trial_found = find_trial_by_identifiers(conn, trial['identifiers'],
+                                            ignore_record_id=record_id)
+
+    if not trial_found:
+        trial_found = find_trial_by_identifiers(conn, trial['identifiers'])
+
+    if trial_found:
+        logger.debug('Trial-id %s was matched via identifiers with register-id %s',
+                     trial_found['id'], record_id)
+        write_log(trial_found['id'], record_id, 'identifier')
+    elif source_id and trial.get('trial_public_title'):
+        trial_temp = conn['database']['trials'].find_one(
+                            public_title=trial.get('trial_public_title'))
+        if trial_temp and trial_temp.get('source_id') != source_id:
```

Comment:
Consider the database as:

| id | public_title | source_id |
| --- | --- | --- |
| A | "A trial" | euctr |
| B | "A trial" | nct |

And a record with `{ public_title: "A trial", source_id: "euctr" }`. If we keep the code as is with the `.find_one()`, it would return the first trial (id: `A`), which would then be ignored as it has the same `source_id` as the record (i.e. `euctr`). The problem is that you're filtering the trials in two separate steps, one querying just for `public_title`, and with the results filter by `source_id`. One solution is to do both filters at once with the SQL query I suggested on https://github.com/opentrials/processors/pull/115#discussion_r103476928.

This is one issue. For the other issue, consider the database as:

| id | public_title | source_id |
| --- | --- | --- |
| A | "A trial" | euctr |
| B | "A trial" | nct |
| C | "A trial" | pfizer |

And the same record `{ public_title: "A trial", source_id: "euctr" }`. If we do a query for `public_title = 'A trial' AND source_id != 'euctr'`, we would get 2 results with sources `nct` and `pfizer`. How should we handle it?

We currently have a priority list implemented on https://github.com/opentrials/processors/blob/master/processors/base/writers/trial.py#L46-L61 to decide which source would become the "primary" trial. This should be replicated here so, in this example, we would pick the trial from `nct` instead of the one from `pfizer`. This would require some refactoring, though, possibly extracting this priority list to a single place (maybe https://github.com/opentrials/processors/blob/master/processors/base/config.py) and reusing it here and in the trial writer.

However, as this is a corner case and we don't have much more time to work on this PR, I'm happy to simply use the first trial from a different source (i.e. `SELECT * FROM trials WHERE public_title = :public_title AND source_id != :source_id LIMIT 1`). We can create an issue for this case and fix it later.